lib/ansible/module_utils/api.py

```diff
@@ -91,26 +92,57 @@ def ratelimited(*args, **kwargs):
     return wrapper
 
 
-def retry(retries=None, retry_pause=1):
-    """Retry decorator"""
-    def wrapper(f):
+def _retry_limit_exceeded(retries):
+    def function_wrapper(function):
+        def raise_error(*args, **kwargs):
+            raise Exception("Retry limit exceeded: %d" % retries)
+        return raise_error
+    return function_wrapper
 
-        def retried(*args, **kwargs):
-            retry_count = 0
-            if retries is not None:
-                ret = None
-                while True:
-                    retry_count += 1
-                    if retry_count >= retries:
-                        raise Exception("Retry limit exceeded: %d" % retries)
-                    try:
-                        ret = f(*args, **kwargs)
-                    except Exception:
-                        pass
-                    if ret:
-                        break
-                    time.sleep(retry_pause)
-                return ret
-
-        return retried
-    return wrapper
+
+def retry(retries=None, retry_pause=1):
+    """Retry decorator.
+    This assumes the function has already been called, so if retries == 1, the maximum attempts
+    have been reached. If retries is None, the function is a no-op.
+    """
+    if retries is None:
+        return lambda function: lambda *args, **kwargs: None
+    if retries <= 1:
+        return _retry_limit_exceeded(retries)
+    try:
+        backoff_delays = [retry_pause for delay in range(0, retries)]
+        return retry_wrapper(backoff_iterator=backoff_delays, retry_condition=lambda x: True)
+    except Exception:
+        raise Exception("Retry limit exceeded: %d" % retries)
+
+
+def jittered_backoff_generator(retries=10, delay=3, max_delay=60):
```

Comment:
This probably shouldn't be a noun and say what it does right in the name:
```suggestion
def emit_jittered_backoff_delays(retries=10, delay=3, max_delay=60):
```
Communicating more specific information in a function name makes it easier to reason about the calling code.

Reply:
"backoff generator" means what it does to me, but I can rename it.

Reply:
There's a convention to call things that do stuff with verbs hinting what the action is while things that just store some state (like constants) go by nouns.

Just compare:
```python
do_something = 42
def do_something():
    print(42)
def a_car():
    drive()
a_car = Car()
a_car.drive()
```

Reply:
`generator` is descriptive to me because I know it will yield values. 'emit' is an unknown, I'd actually have to look at the code to know what that means. But I don't have a problem with renaming this.

Reply:
How about `generate_jittered_backoff()` then?

Reply:
I like that :+1: